File: Valid Parentheses/valid_parentheses.py

def is_valid(s: str) -> bool:
    stack = []
    parenthesis = {
        "(": ")",
        "{": "}",
        "[": "]"
    }
    for i in s:
        if i in parenthesis:
            stack.append(i)
        else:
            if not stack:
                return False
            if parenthesis[stack.pop()] != i:
                return False
    return len(stack) == 0


print(is_valid("()"))
print(is_valid("()[]{}"))
print(is_valid("(]"))

File: Valid Parentheses/valid_parentheses.go
package main

import "fmt"

func main() {
	fmt.Println(isValid("()[]{}"))
}

func isValid(s string) bool {
	parenthesis := map[rune]rune{'(': ')', '[': ']', '{': '}'}
	stack := []rune{}
	for _, p := range s {
		_, openP := parenthesis[p]
		if openP {
			stack = append(stack, p)
		} else {
			if len(stack) == 0 {
				return false
			}
			var topOfStack rune
			topOfStack, stack = stack[len(stack)-1], stack[:len(stack)-1]
			if parenthesis[topOfStack] != p {
				return false
			}
		}
	}
	return len(stack) == 0
}
